### PR TITLE
Move all job PROPERTIES to Jenkins Files for Build and Pipeline jobs

### PR DESCRIPTION
**Pull request by AdamBrousseau**
- Add Build Discarder
  - Retrieve values from Variable file
- Add job Parameters
  - Default values for Vendor Params must be set globally
    even if they are blank
- No Trigger for Build and Pipeline jobs
- TODO add trigger for PR builds

[skip ci]
Issue #1017
Signed-off-by: Adam Brousseau <adam.brousseau88@gmail.com>

Depends ~#2093~

### CODE REVIEW

File: buildenv/jenkins/jobs/builds/Build-Test-Any-Platform [hunk 1]
```diff
@@ -20,6 +20,16 @@
...
 
+// Discard old builds
+NUM_ARTIFACTS = 5
+NUM_BUILDS = 10
+DAYS_BUILDS = 60
+if (params.ghprbPullId) {
+    // Keep last 50 of each PullRequest build
+    NUM_BUILDS = 50
+}
+properties([buildDiscarder(logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: NUM_ARTIFACTS, daysToKeepStr: DAYS_BUILDS, numToKeepStr: NUM_BUILDS))])
```
Comment (vsebe) [inline buildenv/jenkins/jobs/builds/Build-Test-Any-Platform]: Missing strategy?

Comment (vsebe) [reply buildenv/jenkins/jobs/builds/Build-Test-Any-Platform]: looks not to be mandatory

Comment (vsebe) [inline buildenv/jenkins/jobs/builds/Build-Test-Any-Platform]: If you move this code into a fuction in variables-functions then you avoid duplicating the code.

File: buildenv/jenkins/jobs/builds/Build-Test-Any-Platform [hunk 2]
```diff
@@ -20,6 +20,16 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
 
+// Discard old builds
+NUM_ARTIFACTS = 5
```
Comment (vsebe) [inline buildenv/jenkins/jobs/builds/Build-Test-Any-Platform]: move `NUM_ARTIFACTS, NUM_BUILDS , DAYS_BUILDS` to variables file

**Comment by AdamBrousseau** [comment]
jenkins compile win jdk10

**Comment by AdamBrousseau** [comment]
jenkins compile win jdk10

**Comment by AdamBrousseau** [comment]
jenkins compile win jdk10

**Comment by AdamBrousseau** [comment]
jenkins compile win jdk10

**Comment by AdamBrousseau** [comment]
jenkins compile win jdk10

**Comment by AdamBrousseau** [comment]
jenkins compile win jdk10

**Comment by AdamBrousseau** [comment]
Unfortunately having the `properties` step will overwrite the manually configured trigger property in the job even if that section isn't defined in the pipeline. I.e. We are trying to add build discarder but since `trigger` is also a property and its not defined in the code, the PRs trigger section will get obliterated upon the first run of the build. 

This [issue comment](https://issues.jenkins-ci.org/browse/JENKINS-42785?focusedCommentId=293491&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-293491) and the next 3 comments sum it up nicely.

See [pipeline-syntax](https://ci.eclipse.org/openj9/view/Pull%20Requests/job/PullRequest-Compile-JDK10-win_x86-64_cmprssptrs-OpenJ9/pipeline-syntax/) `Sample Step: properties` to see all the properties you can define. We currently use parameters, triggers, and build discarder. Since we set a different SPEC & PLATFORM parameter on every job, I don't this this will work as is.

We might be able to define everything in the jenkinsfiles. We would have to have all the parameters and all the trigger regexes and the user whitelist in the variable file, which is a good thing. Upon initial setup you could leave all the config blank, and we could have an `if` in the pipeline that first checked for `PLATFORM` and `SPEC`, if these were undefined we could [parse the job name](#1396) to set them up. Then all the rest of the properties could be determined afterwards.

I will change this PR to WIP and update the title to reflect the larger task.

**Comment by AdamBrousseau** [comment]
jenkins compile xlinux jdk10

**Comment by AdamBrousseau** [comment]
@vsebe for initial review.

**Comment by AdamBrousseau** [comment]
jenkins compile xlinux jdk10

**Comment by AdamBrousseau** [comment]
Jenkins copyright check

File: buildenv/jenkins/common/variables-functions [hunk 3]
```diff
@@ -51,29 +51,34 @@ def parse_variables_file(){
...
+    VARIABLE_FILE = params.VARIABLE_FILE
+    VENDOR_BRANCH = params.VENDOR_BRANCH
+    VENDOR_REPO = params.VENDOR_REPO
+    VENDOR_CREDENTIALS_ID = params.VENDOR_CREDENTIALS_ID
+    if (VARIABLE_FILE && VENDOR_REPO && VENDOR_BRANCH) {
+        echo "VARIABLE_FILE:'${VARIABLE_FILE}'"
+        echo "VENDOR_REPO:'${VENDOR_REPO}'"
+        echo "VENDOR_BRANCH:'${VENDOR_BRANCH}'"
+        echo "VENDOR_CREDENTIALS_ID:'${VENDOR_CREDENTIALS_ID}'"
+        if (VENDOR_CREDENTIALS_ID != '') {
```
Comment (vsebe) [inline buildenv/jenkins/common/variables-functions]: ` if (VENDOR_CREDENTIALS_ID != '')` replace with  `if (VENDOR_CREDENTIALS_ID)`

File: buildenv/jenkins/common/variables-functions [hunk 4]
```diff
@@ -421,23 +426,166 @@ def set_job_names() {
...
+        PARAMETERS.add(choice(name: 'SDK_VERSION', choices: ["${SDK_VERSION}"], description: ''))
+        PARAMETERS.add(choice(name: 'PLATFORM', choices: ["${PLATFORM}"], description: ''))
+
+        // Pipeline or build
+        if (JOB_TYPE == 'pipeline' || JOB_TYPE == 'build') {
+            add_string_params_no_default(['OPENJDK_REPO', 'OPENJDK_BRANCH', 'OPENJDK_SHA', 'OPENJ9_REPO', 'OPENJ9_BRANCH', 'OPENJ9_SHA', 'OMR_REPO', 'OMR_BRANCH', 'OMR_SHA'])
+            // PLEASE NOTE: The following 4 DEFAULT variables must be configured in the Jenkins Global Config (even if they are set to blank).
+            // This allows us to know what the default values are without being told explicitly.
+            // These DEFAULT values are typically constant per Jenkins Server.
+            PARAMETERS.add(string(name: 'VARIABLE_FILE', defaultValue: "${VARIABLE_FILE_DEFAULT}", description: '', trim: true))
```
Comment (vsebe) [inline buildenv/jenkins/common/variables-functions]: add checks for the *_DEFAULT variables, e.g 
```
if (VARIABLE_FILE_DEFAULT) {
  PARAMETERS.add(string(name: 'VARIABLE_FILE', defaultValue: "${VARIABLE_FILE_DEFAULT}", description: '', trim: true))
}
```

**Comment by AdamBrousseau** [comment]
@smlambert for final review

File: buildenv/jenkins/common/variables-functions [hunk 5]
```diff
@@ -444,16 +449,174 @@ def set_job_names() {
...
+    *       - Ideally get values from variable file
+    *   - Trigger
+    *       - Build/Pipeline - no trigger
+    *       - PR: GH Trigger
+    ****************************/
+
+    if (JOB_TYPE == "build" || JOB_TYPE == "pipeline") {
+        // Start with just supporting Build and Pipeline jobs
+        // PR jobs will need some playing to get the trigger figured out
+        // Also regaring PR builds, see OpenJ9 Issue #2728
```
Comment (smlambert) [inline buildenv/jenkins/common/variables-functions]: spelling regaring -> regarding

Comment (AdamBrousseau) [reply buildenv/jenkins/common/variables-functions]: fixed

**Pull request merged**